### PR TITLE
Refactor: Preparation for a Parquet WAL

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:
Refactors in preparation for laying a parquet wal. Create and OpenWALBlock can likely be cleaned up. Perhaps work for the next PR. It includes:

- Make AppendBlock an interface and rename to WALBlock
- Moved existing AppendBlock code into v2
- Move WALBlock creation into the encoding and implement BackendBlock
- Encoding now receives a generic *tempopb.Trace iterator to create a new Backend Block. It attempts to cast this to an internal iterator that is more performant.
- Moved a bunch of common types into v2
- Hardcoded the creation of v2 WALBlocks. Next PR will allow the creation of vParquet WALBlocks
- Moved localBlock into the /modules/ingester package

Includes all changes in #1745 and extends.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: tempodb/wal/wal.go [hunk 1]
```diff
@@ -97,13 +82,19 @@ func New(c *Config) (*WAL, error) {
...
 // RescanBlocks returns a slice of append blocks from the wal folder
-func (w *WAL) RescanBlocks(fn RangeFunc, additionalStartSlack time.Duration, log log.Logger) ([]*AppendBlock, error) {
+func (w *WAL) RescanBlocks(additionalStartSlack time.Duration, log log.Logger) ([]common.WALBlock, error) {
 	files, err := os.ReadDir(w.c.Filepath)
 	if err != nil {
 		return nil, err
 	}
 
-	blocks := make([]*AppendBlock, 0, len(files))
+	// todo: rescan blocks will need to detect if this is a vParquet or v2 wal file and choose the appropriate encoding
```
Comment (mdisibio) [inline tempodb/wal/wal.go:91]: I think ultimately we will have to decide whether the wal files are owned by tempodb/wal or the encoding and that determines what we do here.  This is good for now as we iterate.

**Review by mdisibio — APPROVED** [review]
LGTM, good cleanup and looks like it should be a lot easier to integrate the parquet wal.  Left a few comments/questions, nothing blocking.

File: tempodb/encoding/v2/create_block.go [hunk 2]
```diff
@@ -12,17 +13,40 @@ import (
...
+	bytesIterator, isBytesIterator := i.(BytesIterator)
+
+	dec, err := model.NewSegmentDecoder(meta.DataEncoding)
+	if err != nil {
+		return nil, fmt.Errorf("error creating segment decoder: %w", err)
+	}
+
+	next := func(ctx context.Context) (common.ID, []byte, error) {
+		// if this is one of our iterators we are in luck. this is quite fast
+		if isBytesIterator {
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply tempodb/encoding/v2/create_block.go:33]: yikes. duh. that's why i made a next var.

File: tempodb/encoding/versioned.go [hunk 3]
```diff
@@ -32,10 +33,16 @@ type VersionedEncoding interface {
...
 	// * EndTime
 	// * TotalObjects
-	CreateBlock(ctx context.Context, cfg *common.BlockConfig, meta *backend.BlockMeta, i common.Iterator, dec model.ObjectDecoder, r backend.Reader, to backend.Writer) (*backend.BlockMeta, error)
+	CreateBlock(ctx context.Context, cfg *common.BlockConfig, meta *backend.BlockMeta, i common.Iterator, r backend.Reader, to backend.Writer) (*backend.BlockMeta, error)
 
 	// CopyBlock from one backend to another.
 	CopyBlock(ctx context.Context, meta *backend.BlockMeta, from backend.Reader, to backend.Writer) error
+
+	// OpenWALBlock opens an existing appendable block for the WAL
+	OpenWALBlock(filename string, path string, ingestionSlack time.Duration, additionalStartSlack time.Duration) (common.WALBlock, error, error)
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply tempodb/encoding/versioned.go:42]: Good points. I did intend to dig into the Open/Create WALBlock functions when I made the Parquet WAL to see if i could simplify.

File: tempodb/tempodb.go [hunk 4]
```diff
@@ -203,28 +202,26 @@ func (rw *readerWriter) WriteBlock(ctx context.Context, c WriteableBlock) error
...
 	// The destination block format:
 	vers, err := encoding.FromVersion(rw.cfg.Block.Version)
 	if err != nil {
 		return nil, err
 	}
 
-	dec := model.MustNewObjectDecoder(block.Meta().DataEncoding)
-
-	iter, err := block.Iterator(combiner)
+	iter, err := block.Iterator()
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply tempodb/tempodb.go:219]: It's closed in the v2 CreateBlock, but not the vParquet CreateBlock. :P

Agree that this is the correct place for it.

File: tempodb/wal/wal.go (hunk 1 above)
Comment (joe-elliott) [reply tempodb/wal/wal.go:91]: I like having a wal object that is responsible for creating and managing the wal blocks. Agree we can revisit in the future if it gets too thin.

**Pull request merged**